Signal/src/ViewControllers/MessageMetadataViewController.swift

```diff
@@ -0,0 +1,393 @@
+            return
+                NSLocalizedString("MESSAGE_STATUS_SENT",
+                                  comment:"message footer for sent messages")
+        } else if message.hasAttachments() {
+            return NSLocalizedString("MESSAGE_STATUS_UPLOADING",
+                                     comment:"message footer while attachment is uploading")
+        } else {
+            assert(message.messageState == .attemptingOut)
+
+            return NSLocalizedString("MESSAGE_STATUS_SENDING",
+                                     comment:"message status while message is sending.")
+        }
+    }
+
+    private func nameLabel(text: String) -> UILabel {
+        let label = UILabel()
+        label.textColor = UIColor.black
+        label.font = UIFont.ows_mediumFont(withSize:14)
+        label.text = text
+        label.setContentHuggingHorizontalHigh()
+        return label
+    }
+
+    private func valueLabel(text: String) -> UILabel {
+        let label = UILabel()
+        label.textColor = UIColor.black
+        label.font = UIFont.ows_regularFont(withSize:14)
+        label.text = text
+        label.setContentHuggingHorizontalLow()
+        return label
+    }
+
+    private func valueRow(name: String, value: String, subtitle: String = "") -> UIView {
+        let row = UIView.container()
+        let nameLabel = self.nameLabel(text:name)
+        let valueLabel = self.valueLabel(text:value)
+        row.addSubview(nameLabel)
+        row.addSubview(valueLabel)
+        nameLabel.autoPinLeadingToSuperView()
+        valueLabel.autoPinTrailingToSuperView()
+        valueLabel.autoPinLeading(toTrailingOf:nameLabel, margin: 10)
+        nameLabel.autoPinEdge(toSuperviewEdge:.top)
+        valueLabel.autoPinEdge(toSuperviewEdge:.top)
+
+        if subtitle.characters.count > 0 {
+            let subtitleLabel = self.valueLabel(text:subtitle)
+            subtitleLabel.textColor = UIColor.ows_darkGray()
+            row.addSubview(subtitleLabel)
+            subtitleLabel.autoPinTrailingToSuperView()
+            subtitleLabel.autoPinLeading(toTrailingOf:nameLabel, margin: 10)
+            subtitleLabel.autoPinEdge(.top, to:.bottom, of:valueLabel, withOffset:1)
+            subtitleLabel.autoPinEdge(toSuperviewEdge:.bottom)
+        } else if value.characters.count > 0 {
+            valueLabel.autoPinEdge(toSuperviewEdge:.bottom)
+        } else {
+            nameLabel.autoPinEdge(toSuperviewEdge:.bottom)
+        }
+
+        return row
+    }
+
+    // MARK: - Actions
+
+    func shareButtonPressed() {
+        if let messageBody = messageBody {
+            UIPasteboard.general.string = messageBody
+            return
+        }
+
+        guard let attachmentStream = attachmentStream else {
+            Logger.error("\(TAG) Message has neither attachment nor message body.")
+            return
+        }
+        AttachmentSharing.showShareUI(forAttachment:attachmentStream)
+    }
+
+    func copyToPasteboard() {
```

Comment:
Is this called anywhere?

Reply:
Not yet, but I was going to do "long-press to copy via menu controller"?